content/posts/flp-inside-out.md

```diff
@@ -0,0 +1,78 @@
+---
+layout: post
+title: FLP Inside Out
+date: 2021-01-26
+author: Tomek Masternak, Szymon Pobiega
+draft: false
+---
+
+In 50 BC an army of Scipio allied by Juba I king of Numidia, surrounded Cezars troops stationing near Thapsus in northern Africa. Their plan was to attack the usurpator in a coordinated fashion so that he could not hold off the two armies combined. Cezar made the first move though and didn't give his opponents a chance to put the plan into action.[^1] 
+
+If only Scipio and Juba had a chance to read the FLP paper would they realize the plan could not work at all! 
+
+## Understanding FLP
+
+The tongue in cheek introduction to this post tries to make the point that FLP is easy to get wrong (and to our experience has been) both at the level of the proof as well as in terms of practical consequences. This post adds some commentary on both topics which can hopefully make the understanding of the result a bit more intuitive.
+
+## The proof
+
+TODO: not showing everyting but only part that are missing from the paper and a great write-up on paper-trail.
+
+### Model strength
+
+TODO: model strenght - the model is very strong i.e. I/O Automata can represent programs that are impossible e.g. they solve the Halting problem [ref to the yale ds notes]. Next the netowrking model is super strong in fact atomic bradcast assumption means that at the newtork layer the consensus is solved (the author could not know about that though as the reduction was to proved at the publication time). 
+
+### Lemma 1
+
+> Suppose that from some configuration C, the schedules sigma1 and sigma2 lead to configurations C1 and C2, respectively. If the sets of processes taking steps in sigma1 and sigma2, respectively, are disjoint, then sigma2 can applied to C1 and sigma1 can be applied to C2, and both lead to the same configuration C3.
+
+This part might be a bit confusiong unless we understand what is the set of processes taking steps in a schedule. Step made by process `p` is defined as a `receive(p, m)` (followed by internal state change and number of `send` operations). So the set of processes taking step in a schedule is a set of all processes `p` that make at least one receive operation in that schedule. It's important to realize that a single step defines not only a process that receives a message but also an exact message that gets received by that process. 
+
+With that in mind Lemma 1 becomes quite intuitive. An internal state of a process and messages that it has sent in a given schedule is fully determined by set of messages it received and order of that receival. If a schedule does not contain process `p` it cannot affect `p` in any way (either by chaning the message set or changing set of messages received).
```

Comment:
`is fully determined by set of messages it received and order of that receival` - I was always wondering where is it actually defined. If we discuss randomness then we can easily represent it as a initial message seeding the RNG but I don't recall seeing it specified explicitly.